Comments: Add author IP address to the response

Copons opened the pull request:
#### Changes proposed in this Pull Request:

Add the author IP address to the response of [`GET /sites/$site/comments/$comment_ID`](https://developer.wordpress.com/docs/api/1.1/get/sites/%24site/comments/%24comment_ID/).

Like the email address, IP address is limited behind a check on the request context (must be set as `edit`) and the requester capabilities (must have `edit_comment`).

#### Testing instructions:

Try the `GET /sites/$site/comments/$comment_ID` endpoint on a Jetpack site here: https://developer.wordpress.com/docs/api/console/

The response's author object must contain an `ip_address` field.
Its value must be `false` if the request `context` parameter is not `edit`. Otherwise it should be an IP address.

#### Note

This is a blocker for the [Comment Management development](https://github.com/Automattic/wp-calypso/projects/32) going on in Calypso.

Copons commented:
Ouch, I didn't expect those whitespace changes.
Do you need a fix or is it ok like this?

zinigor commented:
If the whitespace is the same on the other side, then it's fine , thanks for making a PR!

Copons commented:
@zinigor It looks like those whitespaces are on lines that don't exist on the other side...

eliorivero commented on class.json-api-endpoints.php:
```diff
@@ -1139,7 +1141,8 @@ function get_author( $author, $show_email = false ) {
 
-		$email = $show_email ? (string) $email : false;
+		$email = $show_email_and_ip ? (string) $email : false;
+		$ip_address = $show_email_and_ip ? (string) $ip_address : false;
```
These could be grouped to check the condition once:
```php
if ( $show_email_and_ip ) {
    $email = (string) $email;
    $ip_address = (string) $ip_address;
} else {
    $email = '';
    $ip_address = '';
}
```
Also, why setting them to `false`, a `boolean` type, when they're not meant to be shown? Why not consistently use a `string` type, an empty string?

Copons replied:
I didn't want to change the existing `string|false` behaviour for `$email`, and I just reused it for `$ip_address` since it's shown for the same context/permissions.

eliorivero replied:
Ok yeah, I guess it makes sense not to fix it if it's not broken :D 
The `if` could still be introduced though.

eliorivero commented on class.json-api-endpoints.php:
```diff
@@ -1053,11 +1053,13 @@ function user_can_view_post( $post_id ) {
 	 *
 	 * @param $author user ID, user row, WP_User object, comment row, post row
-	 * @param $show_email output the author's email address?
+	 * @param $show_email_and_ip output the author's email address and IP address?
```
Now that you're writing this, you could fix the missing data type in these two params, object and bool respectively.

Copons replied:
Gladly!
Should I also update it on the other side?

eliorivero replied:
Yeah, that would be cool.

eliorivero commented on class.json-api-endpoints.php:
```diff
@@ -295,11 +295,11 @@ function input( $return_default_values = true, $cast_and_filter = true ) {
+			array( 'method' => 'POST' ),
+			array( 'secure_key' => $secure_key )
 		);
 		if ( 200 !== $response['response']['code'] ) {
```
Should use `wp_remote_retrieve_response_code` function.

Copons replied:
Wait: this is an accidental whitespace change, unrelated to this PR.

eliorivero replied:
Yeah I know, I was just seeing that this could be updated too, but it's not necessary for the scope of this PR.

eliorivero commented on class.json-api-endpoints.php:
```diff
@@ -1053,11 +1053,13 @@ function user_can_view_post( $post_id ) {
 	 */
-	function get_author( $author, $show_email = false ) {
+	function get_author( $author, $show_email_and_ip = false ) {
+		$ip_address = $author->comment_author_IP;
```
Is this always set? Maybe there should be an `isset`/`empty` check here.

Copons replied:
If this is not set, isn't enough the cast to `(string)` we do later to return an empty string?

eliorivero replied:
If the property isn't set this line will throw an PHP notice "undefined property".

Copons commented:
@eliorivero Got around to answer while you were still commenting. :)

Copons commented:
This PR didn't work because I had to whitelist the `ip_address` in the author object in both JP and wpcom sides.
Fixed it and updated after @eliorivero suggestions.

D6877-code is the wpcom equivalent.

Copons commented:
While this is still in review, I'm wondering about something.
We currently return `email` (and eventually `ip_address`) if the request has `context: 'edit'`.

Is there a reason for not returning them on `context: 'display'`?

eliorivero commented:
Tested applying D6877 to wpcom and works fine! 🐑

zinigor commented:
I think we can add these fields to the display context whenever they are needed. For now this is good to go, thank you!

Copons commented:
Thanks!
FYI I've also deployed the twin diff, so we're in sync: D6877-code
